feat: add C `ndarray` API and refactor `blas/ext/base/dsumors`

headlessNode opened the pull request:
## Description

> What is the purpose of this pull request?

This pull request:

-   add C `ndarray` API and refactor/cleanup `blas/ext/base/dsumors`

## Related Issues

> Does this pull request have any related issues?

No. 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

headlessNode committed:
feat: add C ndarray API and refactor

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: missing_dependencies
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsumors/README.md:
```diff
@@ -147,6 +146,123 @@ console.log( v );
+
+#### stdlib_strided_dsumors_ndarray( N, \*X, strideX, offsetX )
+
+Computes the sum of double-precision floating-point strided array elements using ordinary recursivev summation and alternative indexing semantics.
```
```suggestion
Computes the sum of double-precision floating-point strided array elements using ordinary recursive summation and alternative indexing semantics.
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsumors/benchmark/c/benchmark.length.c:
```diff
@@ -107,6 +107,7 @@ static double benchmark( int iterations, int len ) {
 	v = 0.0;
 	t = tic();
 	for ( i = 0; i < iterations; i++ ) {
+		//cppcheck-suppress uninitvar
```
```suggestion
		// cppcheck-suppress uninitvar
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsumors/benchmark/c/benchmark.length.c:
```diff
@@ -120,6 +121,40 @@ static double benchmark( int iterations, int len ) {
+	v = 0.0;
+	t = tic();
+	for ( i = 0; i < iterations; i++ ) {
+		//cppcheck-suppress uninitvar
```
```suggestion
		// cppcheck-suppress uninitvar
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsumors/src/main.c:
```diff
@@ -17,31 +17,47 @@
+	CBLAS_INT ox = stdlib_strided_stride2offset( N, strideX );
+	return API_SUFFIX(stdlib_strided_dsumors_ndarray)( N, X, strideX, ox );
+}
 
 /**
 * Computes the sum of double-precision floating-point strided array elements using ordinary recursive summation.
 *
 * @param N       number of indexed elements
 * @param X       input array
-* @param stride  stride length
+* @param strideX  stride length
+* @param offsetX  starting index
 * @return        output value
 */
```
JSDoc missing `alternative indexing semantics`

headlessNode committed:
refactor: apply code review suggestions

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: passed
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dsumors/README.md:
```diff
@@ -132,11 +130,12 @@ var v = dsumors.ndarray( 4, x, 2, 1 );
 var dsumors = require( '@stdlib/blas/ext/base/dsumors' );
 
-var x = filledarrayBy( 10, 'float64', discreteUniform( 0, 100 ) );
+var x = discreteUniform( 10.0, -100, 100, {
```
```suggestion
var x = discreteUniform( 10, -100, 100, {
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dsumors/examples/index.js:
```diff
@@ -18,11 +18,12 @@
 var dsumors = require( './../lib' );
 
-var x = filledarrayBy( 10, 'float64', discreteUniform( -100.0, 100.0 ) );
+var x = discreteUniform( 10.0, -100, 100, {
```
```suggestion
var x = discreteUniform( 10, -100, 100, {
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dsumors/src/main.c:
```diff
@@ -17,31 +17,47 @@
+/**
+* Computes the sum of double-precision floating-point strided array elements using ordinary recursive summation and alternative indexing semantics.
+*
 * @param N       number of indexed elements
```
@headlessNode The descriptions need to be realigned.

headlessNode replied:
@kgryte Sorry, I'm not sure I understand. What do you mean by "realigned"?

kgryte replied:
@headlessNode They don't all start at the same character column. See L26-29. Compare L40-43 where the descriptions for strideX and offsetX now start one column farther to the right.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dsumors/test/test.dsumors.js:
```diff
@@ -135,14 +135,14 @@ tape( 'the function supports a negative `stride` parameter', function test( t )
 });
 
-tape( 'if provided a `stride` parameter equal to `0`, the function returns the first element', function test( t ) {
+tape( 'if provided a `stride` parameter equal to `0`, the function returns the sum of first element repeated N times', function test( t ) {
```
```suggestion
tape( 'if provided a `stride` parameter equal to `0`, the function returns the sum of the first element repeated N times', function test( t ) {
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dsumors/test/test.dsumors.native.js:
```diff
@@ -226,14 +226,14 @@ tape( 'the function supports a negative `stride` parameter', opts, function test
 });
 
-tape( 'if provided a `stride` parameter equal to `0`, the function returns the first element', opts, function test( t ) {
+tape( 'if provided a `stride` parameter equal to `0`, the function returns the sum of first element repeated N times', opts, function test( t ) {
```
```suggestion
tape( 'if provided a `stride` parameter equal to `0`, the function returns the sum of the first element repeated N times', opts, function test( t ) {
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dsumors/test/test.ndarray.js:
```diff
@@ -135,14 +135,14 @@ tape( 'the function supports a negative `stride` parameter', function test( t )
 });
 
-tape( 'if provided a `stride` parameter equal to `0`, the function returns the first indexed element', function test( t ) {
+tape( 'if provided a `stride` parameter equal to `0`, the function returns the sum of first element repeated N times', function test( t ) {
```
```suggestion
tape( 'if provided a `stride` parameter equal to `0`, the function returns the sum of the first element repeated N times', function test( t ) {
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dsumors/test/test.ndarray.native.js:
```diff
@@ -144,14 +144,14 @@ tape( 'the function supports a negative `stride` parameter', opts, function test
 });
 
-tape( 'if provided a `stride` parameter equal to `0`, the function returns the first indexed element', opts, function test( t ) {
+tape( 'if provided a `stride` parameter equal to `0`, the function returns the sum of first element repeated N times', opts, function test( t ) {
```
```suggestion
tape( 'if provided a `stride` parameter equal to `0`, the function returns the sum of the first element repeated N times', opts, function test( t ) {
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (changes requested):
Other than the description alignment, this LGTM.

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/dsumors/src/main.c:
```diff
@@ -17,31 +17,47 @@
+* @param strideX  stride length
+* @return         output value
+*/
+double stdlib_strided_dsumors( const CBLAS_INT N, const double *X, const CBLAS_INT strideX ) {
+	CBLAS_INT ox = stdlib_strided_stride2offset( N, strideX );
+	return API_SUFFIX(stdlib_strided_dsumors_ndarray)( N, X, strideX, ox );
+}
```
```suggestion
double API_SUFFIX(stdlib_strided_dsumors)( const CBLAS_INT N, const double *X, const CBLAS_INT strideX ) {
	CBLAS_INT ox = stdlib_strided_stride2offset( N, strideX );
	return API_SUFFIX(stdlib_strided_dsumors_ndarray)( N, X, strideX, ox );
}
```

headlessNode committed:
fix: add the missing suffix

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode committed:
docs: apply code review suggestion

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

headlessNode commented:
@kgryte Changes applied.

kgryte reviewed (approved):
LGTM
